aten/src/ATen/native/SpatialAveragePooling.cpp

```diff
@@ -0,0 +1,385 @@
+
+      int64_t i;
+      for(i=0; i<inputWidth*inputHeight; i++)
+        ptr_gi[i] = 0.0;
+
+      for(yy = 0; yy < outputHeight; yy++)
+      {
+        for(xx = 0; xx < outputWidth; xx++)
+        {
+          int64_t hstart = yy * dH - padH;
+          int64_t wstart = xx * dW - padW;
+          int64_t hend = std::min(hstart + kH, inputHeight + padH);
+          int64_t wend = std::min(wstart + kW, inputWidth + padW);
+          int pool_size = (hend - hstart) * (wend - wstart);
+          hstart = std::max(hstart, (int64_t) 0);
+          wstart = std::max(wstart, (int64_t) 0);
+          hend = std::min(hend, inputHeight);
+          wend = std::min(wend, inputWidth);
+
+          scalar_t z = *ptr_gradOutput++;
+
+          int divide_factor;
+          if(count_include_pad)
+            divide_factor = pool_size;
+          else
+            divide_factor = (hend - hstart) * (wend - wstart);
+
+          int64_t kx, ky;
+          for(ky = hstart ; ky < hend; ky++)
+          {
+            for(kx = wstart; kx < wend; kx++)
+              ptr_gradInput[ky*inputWidth + kx] += z/divide_factor;
+          }
+        }
+      }
+    }
+  }
+}
+
+static void spatial_avg_pool2d_backward_cpu_template(
+  Tensor& input,
+  Tensor& gradOutput,
+  Tensor& gradInput,
+  IntList kernel_size,
+  IntList stride_size,
+  IntList pad_size,
+  bool ceil_mode,
+  bool count_include_pad)
+{
+  auto kH = kernel_size[0];
+  auto kW = kernel_size[1];
+  auto dH = stride_size[0];
+  auto dW = stride_size[1];
+  auto padH = pad_size[0];
+  auto padW = pad_size[1];
+  int dimw = 2;
+  int dimh = 1;
+  int dimc = 0;
+  int64_t nbatch = 1;
+  int64_t ndim = 3;
+
+  int64_t inputWidth;
+  int64_t inputHeight;
+  int64_t outputWidth;
+  int64_t outputHeight;
+  int64_t nInputPlane; // number of channels (or colors)
+
+  spatial_avg_pool2d_shapecheck(
+      input, kernel_size, stride_size, pad_size, ceil_mode);
+
+
+  if (input.dim() == 4) {
+    nbatch = input.size(0);
+    dimw++;
+    dimh++;
+    dimc++;
+    ndim = 4;
+  }
+
+  inputWidth = input.size(dimw);
+  inputHeight = input.size(dimh);
+  nInputPlane = input.size(dimc);
+
+  outputWidth = pooling_output_shape<int64_t>(inputWidth, kW, padW, dW, 1, ceil_mode);
+  outputHeight = pooling_output_shape<int64_t>(inputHeight, kH, padH, dH, 1, ceil_mode);
+
+  gradInput.resize_as_(input);
+
+  gradOutput = gradOutput.contiguous();
+  if (!gradInput.is_contiguous())
+    AT_ERROR("gradInput must be contiguous");
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+    gradOutput.scalar_type(), "spatial_avg_pool2d", [&] {
+      spatial_avg_pool2d_backward_cpu_frame<scalar_t>(
+          gradInput,
+          gradOutput,
+          nInputPlane,
+          inputHeight,
+          inputWidth,
+          outputHeight,
+          outputWidth,
+          nbatch,
+          count_include_pad,
+          kernel_size,
+          stride_size,
+          pad_size);
+    });
+}
+
+} // namespace
```

Comment:
I assume you were going to add the wrapper functions here. Can you add the functions that wrap e.g. `spatial_avg_pool2d_out_cpu_template` see for examples `AdaptiveMaxPool2d.cpp` or `FractionalMaxPool2d.cpp`?

Then it's easier to talk about the `output` shape inference.

Reply:
Thanks!